ls1intum/Artemis
# `Plagiarism checks`: Add plagiarism cases for exams

### Checklist
#### General


#### Server
#### Client
- [x] (Indirectly included at the loaded module) I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).

### Motivation and Context


This PR extends the plagiarism features present for course exercises to support exam exercises as well. This allows the instructors to conduct plagiarism checks and to set verdict for exams and allows students to see the plagiarism cases related to their exam submissions.

### Description

New endpoint to list plagiarism cases for an exam for instructors are added. The existing components at client are extended toı support exam exercises.

Additionally, this PR fixes an issue common in both course and exam plagiarism cases where plagiarism case data is not updated when notification for a different plagiarism case is clicked when on the plagiarism case student detail view. Example: When a student clicks the notification for `/courses/1/plagiarism-cases/2` when they are at `/courses/1/plagiarism-cases/1` the page is not updated. 

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam
  - 1 Text Exercise with plagiarism case
  - 1 Modeling Exercise with plagiarism case
  - 1 Programming Exercise with plagiarism case

1. Log in to Artemis
2. Navigate to Course Administration
3. Make two similar submissions with both students.
4. As the instructor, set the end of working time of the exam to now and run plagiarism check from Exam Detail page > Exercise Groups > Detail page of one of the text/modeling/programming exercises > Check Plagiarism page.
5. Notify a student about the plagiarism.
6. With the notified student, ensure they **cannot access any sensitive data** for the related exam or the plagiarism case.
7. With the notified student, check if the exam result page shows the red `Plagiarism Case` button near the affected exercise.
8. As the instructor, set a verdict.
9. With the notified student, check if the plagiarism page shows the verdict (you can navigate there from the received notification or the `Plagiarism Case`/`Resolved Plagiarism Case` button near the affected exercise.
10. Repeat steps above for one of each text/modeling/programming exercise types in an exam.
11. Ensure that exam conduction is not affected and still working as before.
12. Access to repos belonging to programming exercises in exams working as before except when the student is notified for a relevant plagiarism case, in this case the student should be able to access all repos involved in that plagiarism case to compare files.

Navigation from notification bugfix:
1. With a student, navigate to any plagiarism detail that they have access to (or type `/courses/:courseId/plagiarism-cases/:plagiarismCaseId` with some valid id values)
2. At the notification panel, click on a plagiarism case notification for a different plagiarism case.
3. Ensure that the page is updated correctly.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| Post.java | 65% | 86% |
| PlagiarismCaseResource.java | 75% | 95% |
| plagiarism-cases-instructor-view.component.ts | 76.08% | 100% |
| plagiarism-cases.service.ts | 71.42% | 100% |
| plagiarism-case-instructor-detail-view.component.ts | 89% | 98.79% |
| plagiarism-case-student-detail-view.component.ts | 75% | 95.83% |
| exam-detail.component.ts | 75% | 96.22% |
| exam-participation-summary.component.ts | 79.77% | 93.06% |

### Screenshots



![image](https://user-images.githubusercontent.com/18427209/185949979-4bb8060b-b79d-4105-b269-2a44efde1e46.png)

![image](https://user-images.githubusercontent.com/18427209/186154361-801cbff0-5459-4907-912c-c050fcd58dbd.png)

![image](https://user-images.githubusercontent.com/18427209/185950121-96658fd5-320f-458a-9414-60206ea00e9f.png)

![image](https://user-images.githubusercontent.com/18427209/185950574-730ee52f-6e21-4af3-9958-fc7c356b5d18.png)

![image](https://user-images.githubusercontent.com/18427209/185950981-d480a7eb-ebf9-4229-a0c2-b7e96d5d9f9e.png)
